Java API For Spark Streaming

pwendell opened the pull request:
This patch adds a Java API for Spark Streaming. It is modeled after, and fully integrated with, the existing Spark Java API. This merge includes full API functionality for both normal streams and paired streams. It also includes an example in Java and a comprehensive test suite. The test suite improves overall streaming test coverage while also validating the Java <=> Scala type conversions which constitute most of the logic in the API code.

This currently does not support creating a custom DStream or InputDStream in Java. I'm planning to include that in a subsequent merge that is independently reviewed.

It would be great to get some feedback on this in the next few days as I add the additional tests.

pwendell committed:
A very rough, early cut at some Java functionality for Streaming.

pwendell committed:
Adding some initial tests to streaming API.

pwendell committed:
Remving main method from tests

pwendell committed:
Better equality test (thanks Josh)

pwendell committed:
Adding more tests

pwendell committed:
Reduce tests

pwendell committed:
Adding transform and union

pwendell committed:
PairDStream and DStreamLike

pwendell committed:
More pair functions

pwendell committed:
Style cleanup and moving functions

pwendell committed:
Squashing a few TODOs

pwendell committed:
Docs, some tests, and work on
StreamingContext

pwendell committed:
Some docs for the JavaTestUtils

pwendell committed:
Import Cleanup

pwendell committed:
More work on StreamingContext

pwendell committed:
Fixing issue with <Long> types

pwendell committed:
Porting to new Duration class

pwendell committed:
More work on InputStreams

pwendell committed:
Flume example and bug fix

pwendell committed:
Some style cleanup

JoshRosen commented on streaming/src/main/scala/spark/streaming/api/java/JavaPairDStream.scala:
```diff
@@ -0,0 +1,329 @@
+    dstream.combineByKey(createCombiner, mergeValue, mergeCombiners, partitioner)
+  }
+
+  def countByKey(numPartitions: Int): JavaPairDStream[K, JLong] = {
```
We should perform a similar conversion in the regular Spark Java API.  We could do this in a separate pull request, though.

pwendell replied:
I think this may works fine in the normal API, because count() is an action so it directly returns a Long type. In Streaming, count() returns another stream, so it has this issue where Long is actually a parameter of the outer type and the compiler doesn't automate a cast.

JoshRosen replied:
It's still a problem in countByKey() (we already have the fix in countByValue()).

pwendell replied:
oh right, it will be a problem whenever it is a transformation.

pwendell committed:
Two changes:

- Updating countByX() types based on bug fix
- Porting new documentation to Java

pwendell committed:
Making comments consistent w/ Spark style

pwendell committed:
Bug fix

pwendell committed:
Remove AnyRef constraint in updateState

pwendell committed:
Bugfix

pwendell committed:
Stash

pwendell committed:
Bug fix and test cleanup

pwendell committed:
Adding other updateState functions

JoshRosen commented:
Could you add checkpoint(), isCheckpointed(), and getCheckpointFile() to JavaRDDLike and add setCheckpointDir() and checkpointFile() to JavaSparkContext?

pwendell committed:
Adding checkpointing to Java API

pwendell committed:
Adding flatMap

pwendell committed:
Checkpointing in Streaming java API

pwendell committed:
Style cleanup

pwendell committed:
Import fixup

pwendell committed:
Merge branch 'streaming' into streaming-java-api

pwendell committed:
Small doc fix

pwendell committed:
Adding queueStream and some slight refactoring

pwendell committed:
Moving tests to appropriate directory

pwendell committed:
NetworkWordCount example

mateiz commented:
Did you coordinate with TD on merging this? I don't want us to be merging stuff here that hasn't been reviewed by TD, especially as he was in the process of merging the latest changes in master into the streaming branch.
